### ISSUE TITLE
Generated haskell  packages include incorrect dependencies

### ISSUE THREAD

**Issue by MangoIV** [NixOS/nixpkgs#187430]
*It would be nice if anybody here could give me a hint at correct maintainers; perhaps `cabal2nix` maintainers?*

### Describe the bug

In [the generated nix expression for the hackage packages](https://raw.githubusercontent.com/NixOS/nixpkgs/nixos-unstable/pkgs/development/haskell-modules/hackage-packages.nix) `singletons-base_3_1` incorrectly depends on `singletons-th` instead of `singletons-th_3_1` as specified in the `cabal` file.  

This is an issue because `singletons-th` 3.0 won't build with `ghc924`. 

Relevant code:
```nix 
  "singletons-base_3_1" = callPackage
    ({ mkDerivation, base, bytestring, Cabal, deepseq, directory
     , filepath, pretty, process, singletons, singletons-th, tasty
     , tasty-golden, template-haskell, text, th-desugar, turtle
     }:
     mkDerivation {
       pname = "singletons-base";
       version = "3.1";
       sha256 = "1bsfbdbfwiq2nis3r95x06r0q9iypyz4hkvmgvk56bwj6421k7kd";
       revision = "1";
       editedCabalFile = "12p0xzmrkn2bjz6wf9i291bh47s9c0ziz6cvvja65vnzwd8l60ry";
       setupHaskellDepends = [ base Cabal directory filepath ];
       libraryHaskellDepends = [
         base pretty singletons singletons-th template-haskell text
         #                      ^ this should be `singletons-th_3_1`
         th-desugar
       ];
       testHaskellDepends = [
         base bytestring deepseq filepath process tasty tasty-golden text
         turtle
       ];
       description = "A promoted and singled version of the base library";
       license = lib.licenses.bsd3;
       hydraPlatforms = lib.platforms.none;
     }) {};
```

### Expected behavior

- if in a cabal file it's specified that a library depends on a specific version and this version is in `nixpkgs`, make the package depend on that version of the library.

**Comment by cdepillabout** [comment]
@MangoIV The `_X_Y_Z` Haskell packages in Nixpkgs are somewhat of an implementation detail. We don't give any guarantees that they will work.  Although end users are free to use them, we accept PRs fixing packages that don't work, and we occasionally put them to good use in Nixpkgs.  These packages often don't get marked `broken` appropriately.

That said, it would of course be best if all the Haskell packages provided by Nixpkgs work.  The problem is that it is difficult to come up with some sort of system for Haskell packages in Nixpkgs that "just works" in all situations.  If you're not careful, it is very easy to come up with "diamond-dependency" situations.  The current Haskell package set in Nixpkgs, cabal2nix, and hackage2nix work together to mostly avoid diamond-dependencies.

**Comment by MangoIV** [comment]
I understand that it's a hard task to make a "just works solution", however in this case it looks like when generating the hackage nix expression it just doesn't put in the mangled version of the library but instead makes it depend on the non-mangled one which makes it really look like a bug instead of a missing feature.

**Comment by cdepillabout** [comment]
> in this case it looks like when generating the hackage nix expression it just _doesn't put in the mangled version of the library_ but instead makes it depend on the non-mangled one

This is to specifically avoid diamond-dependencies.  I wouldn't call this a "missing feature", but something that was actively engineered to work this way.

When I see a dependency list like:

```nix
({ mkDerivation, base, bytestring, Cabal, deepseq, directory
     , filepath, pretty, process, singletons, singletons-th, tasty
     , tasty-golden, template-haskell, text, th-desugar, turtle
     }:
     mkDerivation {
       pname = "singletons-base";
       version = "3.1";
...
```

I interpret this as saying that `singletons-base` needs `singletons-th`, but it doesn't specify anything about which version of `singletons-th` is necessary.  It is on the user to figure out a coherent package set with versions of packages that all work together.

> makes it really look like a bug instead of a missing feature

If you can come up with nice fix to this this "bug", we'd of course love to use it!  This is something that seems to frequently confuse people, so having something that works more reliably would be great.

I imagine you've seen `haskell.nix`?  Using their Cabal solver to generate a package set might be more what you're looking for?

**Comment by MangoIV** [comment]
> I imagine you've seen haskell.nix? Using their Cabal solver to generate a package set might be more what you're looking for?

Mmmh. I use haskell.nix for work and honestly it's a PITA and I don't wanna work with it for this as I just need a ghc with some packages. 

> I interpret this as saying that singletons-base needs singletons-th, but it doesn't specify anything about which version of singletons-th is necessary. It is on the user to figure out a coherent package set with versions of packages that all work together.

Well, `singletons-base` specifies in its cabal file that it needs `singletons-th`, it just doesn't put the specific version in because it doesn't mangle the name to be `singletons-th_3_1`

**Comment by sternenseemann** [comment]
> I understand that it's a hard task to make a "just works solution", however in this case it looks like when generating the hackage nix expression it just doesn't put in the mangled version of the library but instead makes it depend on the non-mangled one which makes it really look like a bug instead of a missing feature.

I want to add here that really there is no such thing as a correct dependency in this case – the solution maybe looks obvious to humans. The packages are not actually linked, this is just a consequence of their constraints and constraint solving.

nixpkgs haskell package set, however, is built around the idea that we actually inherit one big constraint solving solution which has as many packages working as possible. That is Stackage and we try having as much packages not part of Stackage working as well on a best effort basis. The key then is that `cabal2nix` / `hackage2nix` actually need to have zero knowledge about hackage and available version: It can just mechanistically convert a cabal file to a Nix expression *without any constraint solving*. The solving is offloaded to the curation of the package set, so in the normal case the Nix expression can expect the input `singletons-th` to be the right version.

The underscored packages are there for convenience mostly, so we don't have to manually add some latest version and regenerate the package set if we want to add a hack for some leaf package where we can afford the inconsistency. It is also useful for downstream users of course and alternative package sets, e.g. you could just add the following lines to the 9.2 overlay:

```nix
{
  singletons-base = doDistribute self.singletons-base_3_1;
  singletons-th = doDistribute self.singletons-th_3_1;
}
```

The alternative package sets are really second class citizens in this regard, since they inherit the main package set's constraint solution / curation, which may not work due to the differing core libs / GHC version.

Doing constraint solving for the versioned attributes, however, would not be practical: You would soon run into extra inconsistencies, requiring us to add even more versioned attributes to the package set. This would probably not really stop anywhere, leading to us shipping basically half of hackage instead of just 1-2 versions of every package. That consequence would not be acceptable: The package set is arguably too big already (the Nix parser needs about a second to process `hackage-packages.nix`) and regenerating the package set would also get quite expensive, since we'd need to run the cabal solver for every versioned package at least.

In conclusion, I'd argue that this is neither a missing feature nor a bug, but rather a consequence of the design decisions taken for nixpkgs' haskell infrastructure: If you have a consistent, curated package set (and a inconsistent package repository), it simply won't be possible to get *any given version* of any given package to work.

A probably more meaningful way to combat the actual issue you are having, would be to generate independent package sets for each compiler version (e.g. a package set based on stackage nightly for GHC 9.2), but I don't think the benefits are really justifiable in light of the file sizes we'd need to check into the git repository and the limited maintainer time we have available.

**Comment by MangoIV** [comment]
> nixpkgs haskell package set, however, is built around the idea that we actually inherit one big constraint solving solution which has as many packages working as possible. That is Stackage and we try having as much packages not part of Stackage working as well on a best effort basis. The key then is that cabal2nix / hackage2nix actually need to have zero knowledge about hackage and available version: It can just mechanistically convert a cabal file to a Nix expression without any constraint solving. The solving is offloaded to the curation of the package set, so in the normal case the Nix expression can expect the input singletons-th to be the right version.

This changes the situation completely, thank you for that clarification. 

> The alternative package sets are really second class citizens in this regard, since they inherit the main package set's constraint solution / curation, which may not work due to the differing core libs / GHC version.

this is a pity as some of the packages don't work with ghc > 8107 which is quite outdated at this point. (considering there's the fourth version of ghc92 and ghc941 out)

> A probably more meaningful way to combat the actual issue you are having, would be to generate independent package sets for each compiler version (e.g. a package set based on stackage nightly for GHC 9.2), but I don't think the benefits are really justifiable in light of the file sizes we'd need to check into the git repository and the limited maintainer time we have available.

maybe this is not necessary; it would already help to have the nix expression that generates the packageset; if you really want to depend on this, you can build it yourself? 

I don't see any reason why the nixpkgs hydra would have to build everything that can be theoretically distributed?

**Comment by MangoIV** [comment]
I see that the `dream2nix` guys are working on a better solution for haskell but I haven't tried it out and `haskell.nix` is a pain to work with in my experience. I quite like the way of how nixpkgs handles haskell packages as it conforms with the normal nixpkgs experience and it's easy to use.

**Comment by maralorn** [comment]
@MangoIV Having more tooling to automatically generate different constraint-solving-solutions for the haskell package set in nixpkgs is a frequently desired feature. (At least frequently desired by me.) It would cover quite a few of the use cases, which currently can only be covered by haskell.nix without manual intervention.
I mean something like `haskell.ghc922.packageSetFor "singletons" "3.1"`. This is not impossible to pull off, but there is no one working on this.

**Comment by MangoIV** [comment]
I agree. I would love to help with working on this.
